openc3/lib/openc3/utilities/local_mode.rb

```diff
@@ -0,0 +1,507 @@
+          next unless File.directory?(full_scope_dir)
+          local_plugins[scope_dir] ||= {}
+          Dir.each_child("#{OPENC3_LOCAL_MODE_PATH}/#{scope_dir}") do |plugin_dir|
+            full_folder_path = "#{OPENC3_LOCAL_MODE_PATH}/#{scope_dir}/#{plugin_dir}"
+            next if plugin_dir == "targets_modified" or not File.directory?(full_folder_path)
+            gems, plugin_instance = scan_plugin_dir(full_folder_path)
+            local_plugins[scope_dir][full_folder_path] = {gems: gems, plugin_instance: plugin_instance}
+          end
+        end
+
+        return local_plugins
+      end
+      return {}
+    end
+
+    def self.analyze_local_mode(plugin_name:, scope:)
+      if ENV['OPENC3_LOCAL_MODE'] and Dir.exist?(OPENC3_LOCAL_MODE_PATH)
+        # We already know a plugin with this name doesn't exist in the models
+        # Now need to determine if there is a highly likely candidate that has been
+        # updated, so that we don't do an erroneous extra plugin install
+
+        gem_name = plugin_name.split("__")[0].split('-')[0..-2].join('-')
+
+        local_plugins = scan_local_mode()
+        scope_plugins = local_plugins[scope]
+        if scope_plugins
+          # Scan models for same gem
+          found_models = {}
+          models = OpenC3::PluginModel.all(scope: scope)
+          models.each do |name, details|
+            model_gem_name = name.split("__")[0].split('-')[0..-2].join('-')
+            found_models[name] = details if gem_name == model_gem_name
+          end
+
+          # Definitely new install if no found models
+          return nil if found_models.length == 0
+
+          # Scan local for same gem and try to match pairs
+          found_local_plugins = {}
+          scope_plugins.each do |folder_path, details|
+            gems = details[:gems]
+            plugin_instance = details[:plugin_instance]
+            next if gems.length != 1
+
+            local_gem_name = File.basename(gems[0]).split('-')[0..-2].join('-')
+            if gem_name == local_gem_name
+              # Gems match - Do the names match?
+              data = File.read(plugin_instance)
+              json = JSON.parse(data, :allow_nan => true, :create_additions => true)
+
+              found = false
+              found_models.each do |name, model_details|
+                if json["name"] == name
+                  # Matched pair
+                  found = true
+                  break
+                end
+              end
+
+              if found
+                found_models.delete(json["name"])
+              else
+                found_local_plugins[folder_path] = details
+              end
+            end
+          end
+
+          # At this point we only have unmatched plugins
+
+          # Not a local mode install if no found local plugins
+          return nil if found_local_plugins.length == 0
+
+          # If we have any unmatched models, assume this should match the first
```

Comment:
Really? Why would it match the first? Is there some sort of text matching thing you should be doing?

Reply:
This handles the case with the sentence you were confused about.  See the updated text in my next commit.